### PR TITLE
 Allow internal drivers with no dependencies to be disabled at compile time

### PR DESCRIPTION
**Pull request by tbonfort**
This PR modifies the autotools build system to allow a user to selectively enable or disable some drivers that were previously always enabled and compiled.

Allows for smaller builds e.g when creating docker images

Example configure invocation to build a gdal image use only for accessing COGs (supposing the result is written to a MEM dataset)

```
./configure \
        --disable-all-optional-drivers \
	--with-threads     \
	--with-png=internal \
	--with-jpeg=internal \
	--with-libtiff=internal \
	--with-libz=internal \
	--with-curl       \
	--without-bsb     \
	--without-cfitsio     \
	--without-cryptopp     \
	--without-ecw     \
	--without-expat     \
	--without-fme     \
	--without-freexl     \
	--without-gif     \
	--without-gnm     \
	--without-grass     \
	--without-grib     \
	--without-hdf4     \
	--without-hdf5     \
	--without-idb     \
	--without-ingres     \
	--without-jasper     \
	--without-jp2mrsid         \
	--without-kakadu     \
	--without-libgrass     \
	--without-libkml     \
	--without-mrf     \
	--without-mrsid     \
	--without-mysql     \
	--without-netcdf     \
	--without-odbc     \
	--without-ogdi     \
	--without-openjpeg     \
	--without-pcidsk     \
	--without-pcraster     \
	--without-pcre     \
	--without-perl     \
	--without-pg        \
	--without-python     \
	--without-qhull     \
	--without-sde     \
	--without-sqlite3     \
	--without-webp     \
	--without-xerces     \
	--without-xml2 \
	--without-lerc \
	--without-geos

```

OGR drivers that cannot be disabled and will always be built: generic geojson kml mem mitab vrt

GDAL drivers that cannot be disabled and will always be built: derived gtiff hfa mem raw vrt

### CODE REVIEW

File: gdal/aclocal.m4 [hunk 1]
```diff
@@ -12,44 +12,6 @@
 # PARTICULAR PURPOSE.
 
 m4_ifndef([AC_CONFIG_MACRO_DIRS], [m4_defun([_AM_CONFIG_MACRO_DIRS], [])m4_defun([AC_CONFIG_MACRO_DIRS], [_AM_CONFIG_MACRO_DIRS($@)])])
-# ===========================================================================
```
Comment (rouault) [inline gdal/aclocal.m4]: is this change needed ?

File: gdal/configure [hunk 2]
```diff
@@ -23736,554 +24120,2883 @@ fi
...
+OGRFORMATS_ENABLED="generic geojson kml mem mitab vrt"
+OGRFORMATS_DISABLED=
 
-PG_CONFIG=no
+# Check whether --enable-format-aaigrid was given.
+if test "${enable_format_aaigrid+set}" = set; then :
+  enableval=$enable_format_aaigrid;
+    FORMAT_AAIGRID_ENABLED=$enableval
+    if test "$enableval" != "no"; then :
+  GDALFORMATS_ENABLED="$GDALFORMATS_ENABLED aaigrid"
```
Comment (rouault) [inline gdal/configure]: odd indentation

File: gdal/scripts/genfrmts.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
```
Comment (rouault) [inline gdal/scripts/genfrmts.sh]: will this file be needed after your work ?

File: gdal/aclocal.m4 (hunk 1 above)
Comment (tbonfort) [reply gdal/aclocal.m4]: probably not, I'll check

File: gdal/scripts/genfrmts.sh (hunk 3 above)
Comment (tbonfort) [reply gdal/scripts/genfrmts.sh]: it will be removed

**Comment by tbonfort** [comment]
> Thanks for tackling this. A bit hard to review given the amount of lines changed though
> 
I'm reworking the change to use looping in configure.ac, should cut down the size of the diff
> * I have perhaps missed it, but I didn't see where the OGR Curl drivers were enabled only if curl is enabled itself.

yeah, the gdal and ogr build stretgies are somewhat different. The curl stuff in OGR is (as of now) done by the 
```
ifneq ($(DRIVER_AMIGOCLOUD_ENABLED),no)
￼   SUBDIRS-$(CURL_SETTING) += amigocloud
￼ endif
```
lines in ogrsf_frmts/GnuMakefile. I'm planning on moving that to a more explicit check in the configure script.

> * More fundamentally, the reality of driver interdependency is much more complicated. For example frmts/iso8211 is used by a number of other drivers. This is just an example. I'm not sure if we want to track driver interdependency, or just accept (and warn !) that this is a advanced functionality and users must be ready to have compilation failures if they disable wrong combinations of drivers.

Correct, there are many more. I'd be OK with marking this feature as advanced usage. We can also add some specific checks at configure time, but that would have to be done  case-by-case. The mitab ogr driver and raw gdal ones are specific cases, as they seem to be used in other places than just other drivers, I'll try to dig a bit more

> * More complicated, and perhaps more theoretical drivers: some drivers or algorithms may require, at runtime, another driver (using GDALGetDriverByName("foo")), but this is generally base drivers like MEM, GTiff, etc. Normally the code properly null-checks the result to error out (since the user can exclude at runtime any driver)

I left in mem, vrt and raw drivers for that reason. kml and geojson are also somewhat related to this, and they have been left in. If there are others we can also leave them out of this patch.

> * From a user point of view, I'm not sure the naming --disable-format-foo vs --disable-driver-foo is really needed. Yes this is how things have been traditionnaly divided, bt I'd prefer --disable-driver-foo both for frmts/ and ogr/ drivers (ok, some have the same name, but in that case we can disable/enable both)

given that the build setup for gdal and ogr are quite different, handling the disabling of both ogr and gdal drivers with the same name is going to prove a bit tricky. I'll see what I can do.

> * Still from user point of view, I believe we need a --disable-all flag to be used in combination with a few --enable-driver-foo invokation. I guess this is how this functionnality will be used by 99% of people.

Agreed. Not sure how to handle this in practice, but I'll have a look also

> * We'll probably need to add in CONTRIBUTING.md guidelines of what is needed now when adding a new driver.

Hopefully this will just sum up to adding the name of the directory in configure.ac

**Comment by tbonfort** [comment]
I've heavily refactored my first draft to take into account your remarks:
- use ```--disable-driver-*``` for gdal and ogr
- remove bloat
- check for ogr and gdal drivers with same name

Still remaining to do:
- figure out weird indentation in generated ```configure```

Still remaining to do, but I'd prefer to add this in a following pull request
- support for ```--disable-all```
- look into refactoring "raw" and "mitab" drivers so they can be disabled
- move logic of enabling/disabling optional drivers that require a dependency from the Makefiles to the configure script

**Comment by tbonfort** [comment]
@rouault I've fixed the indentation part, however I fail to understand why the (only python?) tests are failing on some platforms

File: gdal/configure.ac [hunk 4]
```diff
@@ -1347,6 +1347,133 @@ fi
...
+    cur_driver_enabled=no
+  fi
+fi
+])
+
+
+
+m4_foreach_w([frmt], INTERNAL_FORMATS,
+[
+AC_ARG_ENABLE(driver-frmt, AS_HELP_STRING([--disable-driver-frmt], [disable frmt driver support (enabled by default)]))dnl
```
Comment (rouault) [inline gdal/configure.ac]: dnl

File: gdal/configure.ac [hunk 5]
```diff
@@ -1347,6 +1347,133 @@ fi
...
+  m4_join([_],[INTERNAL_FORMAT],frmt,[ENABLED])=yes
+else
+  GDALFORMATS_DISABLED="$GDALFORMATS_DISABLED frmt"
+  m4_join([_],[INTERNAL_FORMAT],frmt,[ENABLED])=no
+fi
+])
+
+m4_foreach_w([frmt],INTERNAL_OPT_FORMATS,
+[
+AC_ARG_ENABLE(driver-frmt, AS_HELP_STRING([--disable-driver-frmt], [disable frmt format support (enabled by default, requires dependency)]))dnl
```
Comment (rouault) [inline gdal/configure.ac]: dnl

File: gdal/configure.ac [hunk 6]
```diff
@@ -5802,6 +5949,10 @@ LOC_MSG([  MongoDB support:           ${MONGODB_ENABLED}])
 LOC_MSG([  MongoCXX v3 support:       ${MONGOCXXV3_ENABLED}])
 LOC_MSG([  HDFS support:              ${HDFS_ENABLED}])
 LOC_MSG([  userfaultfd support:       ${ENABLE_UFFD}])
+LOC_MSG([  misc. gdal formats:       ${GDALFORMATS_ENABLED}])
```
Comment (rouault) [inline gdal/configure.ac]: missing space character for alignment

**Comment by rouault** [comment]
https://trac.osgeo.org/gdal/wiki/BuildingOnUnixWithMinimizedDrivers will need to be updated

File: gdal/configure.ac (hunk 6 above)
Comment (rouault) [reply gdal/configure.ac]: ignore this remark. Just a diff artifact. The output of https://travis-ci.com/OSGeo/gdal/jobs/174719812 looks good

Comment (tbonfort) [reply gdal/configure.ac]: actually it's not a diff artifact, I removed a space as GDAL_FORMATS_ENABLED will start with a space due to how it is constructed

File: gdal/configure.ac [hunk 7]
```diff
@@ -1347,6 +1347,133 @@ fi
...
+AC_DEFUN([INTERNAL_DRIVERS],[aeronavfaa arcgen avc bna cad csv dgn dxf edigeo geoconcept georss gml gmt gpsbabel gpx gtm htf jml mvt ntf openair openfilegdb pgdump rec s57 segukooa segy selafin shape sua svg sxf tiger vdv wasp xplane])dnl
+AC_DEFUN([INTERNAL_CURL_DRIVERS],[amigocloud carto cloudant couchdb csw elastic gft ngw plscenes wfs])dnl
+
+AC_ARG_ENABLE(all-drivers, AS_HELP_STRING([--disable-all-drivers], [disable all internal drivers. Individual drivers can be added with --enable-driver-XXX]))
+if test "x$enable_all_drivers" = "xno"; then :
+  all_drivers_disabled=yes
+else
+  all_drivers_disabled=no
+fi
+dnl m4_foreach_w([frmt], INTERNAL_FORMATS,
```
[the remark this answers is not in the document]
Comment (tbonfort) [reply gdal/configure.ac]: to be removed

File: gdal/configure.ac [hunk 8]
```diff
@@ -1347,6 +1347,133 @@ fi
...
 
+GDALFORMATS_ENABLED=
+GDALFORMATS_DISABLED=
+OGRFORMATS_ENABLED=
+OGRFORMATS_ENABLED_CFLAGS=
+OGRFORMATS_DISABLED=
+
+AC_DEFUN([INTERNAL_FORMATS],[aaigrid adrg aigrid airsar arg blx bmp cals ceos ceos2 coasp cosar ctg dimap dted e00grid elas envisat ers fit gff gsg gxf hf2 idrisi ignfheightasciigrid ilwis ingr iris iso8211 jaxapalsar jdem kmlsuperoverlay l1b leveller map msgn ngsgeoid nitf northwood pds prf r rmf rs2 safe saga sdts sentinel2 sgi sigdem srtmhgt terragen til tsx usgsdem xpm xyz zmap])
+AC_DEFUN([INTERNAL_OPT_FORMATS],[eeda plmosaic ozi pdf rda rik wcs wms wmts])
+AC_DEFUN([INTERNAL_DRIVERS],[aeronavfaa arcgen avc bna cad csv dgn dxf edigeo geoconcept georss gml gmt gpsbabel gpx gtm htf jml mvt ntf openair openfilegdb pgdump rec s57 segukooa segy selafin shape sua svg sxf tiger vdv wasp xplane])dnl
```
[the remark this answers is not in the document]
Comment (tbonfort) [reply gdal/configure.ac]: all the dnl parts prevent empty lines from being emitted in the generated configure. I have no strong preference...

[the remark this answers is not in the document]
Comment (rouault) [reply gdal/configure.ac]: oh ok, so was intended. keep it then

**Comment by tbonfort** [comment]
PR squashed and ready for merge. --disable-all-drivers has been renamed to --disable-all-optional-drivers.
